nixos/modules/services/security/step-ca.nix

```diff
@@ -0,0 +1,113 @@
+      };
+      keyFile = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = "The path of an existing key file of the root certificate authority.";
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.step-ca = {
+      wantedBy = [ "multi-user.target" ]; # todo fixme this should be like acme
+      after = [ "network.target" ];
+      path = [ pkgs.coreutils pkgs.step-ca pkgs.step-cli ];
+      environment = { STEPPATH = "/var/lib/${cfg.stateDir}"; HOME = "/var/lib/${cfg.stateDir}"; };
+      script = ''
+        if [ ! -f "$STATE_DIRECTORY/ca-is-init" ]; then
+          step ca init --ssh --name ${cfg.name} --dns ${cfg.hostname} --address ${cfg.address}:${toString cfg.port} --provisioner ${cfg.provisioner} --password-file ${cfg.passwordFile} \
+            ${lib.optionalString (cfg.rootFile != null) " --root=${cfg.rootFile} "} ${lib.optionalString (cfg.keyFile != null) " --key=${cfg.keyFile} "}
+          ${lib.optionalString (cfg.acmeProvisioner != null) "step ca provisioner add ${cfg.acmeProvisioner} --type ACME"}
+          touch $STATE_DIRECTORY/ca-is-init
+        fi
+        step-ca $STEPPATH/config/ca.json --password-file ${cfg.passwordFile}
+      '';
+      serviceConfig = {
+        Type = "simple";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        Restart = "on-failure";
+        StateDirectory = cfg.stateDir;
+
+        ProtectProc = "invisible";
+        ProcSubset = "pid";
+        CapabilityBoundingSet = "";
+        NoNewPrivileges = "true";
+        ProtectSystem = "full";
+        ProtectHome = "true";
+        PrivateTmp = "true";
+        PrivateDevices = "true";
+        PrivateUsers = "true";
+        ProtectHostname = "true";
+        ProtectClock = "true";
+        ProtectKernelTunables = "true";
+        ProtectKernelModules = "true";
+        ProtectKernelLogs = "true";
+        ProtectControlGroups = "true";
+        RestrictAddressFamilies = "AF_INET AF_INET6";
+        RestrictNamespaces = "true";
+        LockPersonality = "true";
+        RestrictRealtime = "true";
+        RestrictSUIDSGID= "true";
+        RemoveIPC = "true";
+        SystemCallFilter = "~@clock @debug @module @mount @raw-io @reboot @swap @privileged @resources @cpu-emulation @obsolete"; # maybe whitelist
+        SystemCallArchitectures = "native";
+        MemoryDenyWriteExecute = "true";
+	      DynamicUser = true;
```

Comment:
Some of the options here looks a bit frightening to be set as default. Especially `SystemCallFilter`, because it has high potential to subtly break things. I don't think we should set that particular option unless either upstream projects actively support it, or we have high confidence that things wouldn't break even with future releases. It's easy for users to override systemd service settings in their `configuration.nix`, so I don't think we need to be too aggressive with the defaults.

Reply:
Other options of concern are:

<dl>
<dt>ProcSubset</dt>
<dd>
From <a href="https://www.freedesktop.org/software/systemd/man/systemd.exec.html">systemd.exec(5)</a>:
<br>
<blockquote>this option is useful only in a few, specific cases, and is not suitable for most non-trivial programs.</blockquote>
</dd>
<dt>CapabilityBoundingSet</dt>
<dd>Similar concerns with <code>SystemCallFilter</code>.</dd>
<dt>ProtectHostname</dt>
<dd>
From <a href="https://www.freedesktop.org/software/systemd/man/systemd.exec.html">systemd.exec(5)</a>:
<blockquote>Note that when this option is enabled for a service hostname changes no longer propagate from the system into the service, it is hence not suitable for services that need to take notice of system hostname changes dynamically.</blockquote>

I have the feeling this might cause problems in some cases with the ACME API server.
</dd>
<dt>RestrictAddressFamilies</dt>
<dd>I feel slightly uneasy about this one too, but maybe we can keep this one. Either way, I think we should include  <code>AF_UNIX</code> in the allowed protocols because of its wide use. Its absence might cause problems in future releases, if not now.</dd>
</dl>

Other minor stuff:

<dl>
<dt>ProtectSystem, ProtectHome, PrivateTmp, RestrictSUIDSGID, RemoveIPC</dt>
<dd>These would be enabled when <code>DynamicUser</code> is enabled.</dd>
</dl>

Reply:
Also, it's preferable to use the boolean `true` instead of the string `"true"`.

Reply:
The remaining options looks fine to me.

Reply:
It would be **great** if you could open a PR upstream and contribute a `.service` file. If upstream isn't interested in maintaining a `.service` file, maybe you could expand the service unit they describe [here](https://github.com/smallstep/certificates/blob/647b9b45416b5a103386807a99407f5624d6f077/docs/GETTING_STARTED.md#systemctl).

Why? Having upstream review these hardening options is a big win... because upstream likely knows better than we do. As a bonus, if you create a PR to add a `.service` file and upstream accepts this, we can avoid having to declare a `systemd` service ourselves and simply utilize the service provided by the package.

Reply:
https://github.com/smallstep/docs/blob/master/step-ca/certificate-authority-server-production.mdx#running-step-ca-as-a-daemon seems to contain a more production ready service file. I will try to improve this and if I'm successful I would ask to add this to the official repository / package.

Reply:
Upstream here, checking in! First of all, thank you all for the scrutiny on this file. I've just been locking down our systemd unit file earlier this month, and it's my first foray into writing a unit file. So, your input here is _super_ helpful!

We are happy to maintain a service unit in our documentation.

[Here's the most recent version of our service unit](https://gist.github.com/tashian/ea70829edd568887c6409e28d2f29282) (still pending deployment to our docs website). The PR @mohe2015 just submitted is quite a bit more locked-down, and in some ways I think is a bit _too_ locked down, but I will go through it with an engineer on our end and get that PR merged.

Reply:
> We are happy to maintain a service unit in our documentation.

That is great news. I'm not sure if you're interested, but maintaining a service unit in your git repo is even better from an OS packaging point of view. If you do this we can directly consume your `.service` file saving the copy+paste, automatically staying in sync with upstream.

Thank you very much for taking an interest!

Reply:
Thanks @aanderse, in that case we will host it [here, in the smallstep/certificates repo](https://github.com/smallstep/certificates/blob/82f82d438c50c7a7107f92a3d63b3717d7e219c9/systemd/step-ca.service) (PR is still pending review).

We've incorporated a lot of the changes from [@mohe2015's PR on our docs repo](https://github.com/smallstep/docs/pull/30). It would be nice to use a dynamic user, but it's not obvious to me how to migrate to this (see [my comment on the PR](https://github.com/smallstep/docs/pull/30#issuecomment-768660909)).

Let me know if there's anything else I can do to help! 🤓

Reply:
I'd prefer we don't consume the upstream service unit directly for the following reasons:

- Creating service units through the `systemd` module would allow users to override individual settings. Directly consuming the upstream service unit would disallow this.
- This PR currently uses `systemd.services.step-ca.script` to generate the step-ca configuration on initial run. As a consequence, we'll have to make intrusive changes to the upstream service unit through patching. 

Regarding my second point, we could alternatively create a freeform module to directly configure `ca.json` instead. This would be a huge win for NixOS because it would make the step-ca configuration truly declarative and rollback-able. But unfortunately, configuring `ca.json` directly seems to be less documented than creating the configuration interactively with `step ca init` and is likely to cause confusion.

Reply:
> Creating service units through the systemd module would allow users to override individual settings. Directly consuming the upstream service unit would disallow this.

This is incorrect. NixOS and `systemd` both allow you to override any part of upstream units. `systemd` documentation goes over in extensive detail how this is accomplished and has specific folders for both distro and sysadmin modifications to apply to any unit file. NixOS, of course, makes this even easier.

> This PR currently uses systemd.services.step-ca.script to generate the step-ca configuration on initial run. As a consequence, we'll have to make intrusive changes to the upstream service unit through patching.

Also incorrect... No patching required. A distro can take what they want and with trivial effort replace any other parts. In NixOS this is even easier than other distros.

Reply:
My bad, I completely forgot about `overrides.conf`. It looks like the NixOS `systemd` module would automatically create an `overrides.conf` file if the upstream service unit is included in the package, so forget about what I said earlier.

Reply:
Indeed!

Reply:
> Regarding my second point, we could alternatively create a freeform module to directly configure `ca.json` instead. This would be a huge win for NixOS because it would make the step-ca configuration truly declarative and rollback-able. But unfortunately, configuring `ca.json` directly seems to be less documented than creating the configuration interactively with `step ca init` and is likely to cause confusion.

https://gist.github.com/mohe2015/9ac9529db54c3f2123cf06e5d424887d is a sample file with quickly added provisioners from the help-page (don't worry I won't use it :D).
To me it seems like this would not be easily feasible but I don't know how we could make this configurable properly otherwise. (btw. it seems like name and type combined need to be unique so you can create two provisioners with the same name but different type @tashian can you confirm this?).
An option would probably be to have something like this:
```nix
[
 {
  name = "acme"; # required
  type = "ACME";  # required
 }
 {
  name = "max@smallstep.com";
  type = "JWK";
  ssh = null;
  create = null;
 }
]
```
which will create the following commands at startup:
```bash
# TODO remove all that are in
step ca provisioner list
# but not in the nix config
step ca provisioner remove ${name} --type ${type}

# TODO check that it does not exist yet in
step ca provisioner list
# if it doesn't:
step ca provsioner add ${name} --type ${type} --ca-config=${caConfig} --{key} {optionalString (value != null) value} ...

# so in this case (without the checks)
step ca provisioner add acme --type ACME
step ca provisioner add max@smallstep.com --ca-config ca.json --ssh --create
```

@tashian is it possible to create a CA without any provisioners?

Reply:
We could do the same for `step ca init` so just passing the options through but I don't know if that is useful in that case. Also I don't know how types could be done then (probably copy pasting from the manpage (and converting to nix))

Reply:
@tashian Also is it possible to pass an existing indermediate certificate to `step ca init` or somehow create a CA from existing root and indermediate certificates (so you don't need to send the root key to the server you deploy the CA on)?

Reply:
Usually (not always?) NixOS tries not to destroy state. Destroying state can be a dangerous thing to do and care must be taken. Consider if someone comments a few lines of their `configuration.nix` out - would the destroyed state be trivial to regenerate, and would anything else be lost?

Just a few things to consider.

Reply:
@aanderse Yeah that was one of the reasons I didn't like my idea that much but I forgot to mention that. I don't know of another way of doing this though.

Reply:
Unfortunately I'm not familiar with this program so have limited insight. Sorry 😔

Reply:
The probably formally more correct but likely not practical approach would be to generate all the keys etc. manually and reference them in the config file. So:
```nix
[
  {
    name = "acme";
    type = "ACME";
   # seems to have no data associated with it?
  }
  {
    name = "jwk";
    type = "JWK";
    key = /run/keys/some-jwk-key; # needs to be valid json?
    encryptedKey = /run/keys/the-jwk-key-encrypted; # needs to be valid json?
    # see https://github.com/smallstep/certificates/discussions/427 for a description of how this works
    claims = {
      enableSSHCA = true;
    };
  }
  {
    name = "Google";
    type = "OIDC";
    clientID = "";
    clientSecret = "";
    configurationEndpoint = "";
  }
]
```
If all referenced files need to be json then the generation itself could probably be automated. The problem is that we should probably provide instruction on how to generate the keys then or reference step-cli commands  for that.

Reply:
After reading a brief introduction to this software it seems the intention is to manage configuration of the application through the `step` to.

NixOS, and `systemd` to a certain extent, generally treat configuration as declarative and immutable (the evidence that `systemd` sees things this way is with `ConfigurationDirectory` enforcing `root` ownership of everything in `/etc`).

**Sometimes** there just isn't much room for mixing declarative configuration with application controlled configuration in NixOS. The module author needs to think carefully whether the sysadmin using a module has net benefit from a declarative configuration given the additional knowledge and complexity it could add to working with the application.

If working with the application becomes sufficiently complex from introducing declarative configuration then it seems like treating configuration as application state (as far as NixOS is concerned) is likely a better solution.

While uninformed of this specific software I hope my thoughts provide some value.

Reply:
Yeah thanks @aanderse I think I put my two ideas here now (maybe somebody has something else in mind) and I would need to get some opinion from people who would potentially use this software or @tashian (sorry that you have so much to read through here) who probably has the knowledge of common usage patterns. I personally think the second approach of putting everything into the NixOS configuration is better. I won't be using this in any professional use but rather for my home server services but I need to pregenerate the root-key so I can put it in the trust-store declaratively so I would also think that configuring the provisioners there would make sense.

Reply:
The problem with doing it declaratively is that you can't add any provisioners imperatively so if this is a common usage we really shouldn't do that. @tashian I would like to know what you think about this? Are there common uses where you need to add/remove provisioners pretty quickly. In NixOS it's possible to edit the configuration and redeploy quickly too but it's not imperative and more a manual approach. So you *usually* don't write a script to do this. Although general usage patterns on NixOS are likely quite different to other distributions so it may be fine here even if it wouldn't be for other distributions.
Is there any provisioner that can't really be created declaratively because of some network interactions etc.? (Azure did something strange as the step-ca service failed to started when I added it)

Reply:
First, just following up on some questions here:

> (btw. it seems like name and type combined need to be unique so you can create two provisioners with the same name but different type @tashian can you confirm this?).

The names should be unique, because provisioners can be referenced in the `step` command, using `--provisioner name`.

> @tashian is it possible to create a CA without any provisioners?

Not using `step ca init`—but you could create a `ca.json` with an empty `provisioners` list and the CA will happily start running.

> @tashian Also is it possible to pass an existing intermediate certificate to step ca init or somehow create a CA from existing root and intermediate certificates (so you don't need to send the root key to the server you deploy the CA on)? 

You can import an existing root with `--root` and `--key`.

Alternatively, you could run `step ca init` to set up the directory structure, then bring  your own certs and key. Easiest way to do this is to create a CA independently using `step certificate create` ((see this doc)[https://smallstep.com/docs/step-cli/basic-crypto-operations#run-an-offline-x509-certificate-authority]), then copy the root cert, intermediate cert, and intermediate key in the default locations in `$STEPPATH`, and they will be used by `step-ca`.

> Also is it possible to set different root and intermediate passwords?

Not with `step ca init` directly. We suggest people run `step crypto change-pass $STEPPATH/secrets/intermediate_ca_key` to change the intermediate key password after bootstrapping their PKI.

> Edit: @tashian is it possible to "disable" provisioners or do you have another idea?

There's no config option for disabling a provisioner.

> Is there any provisioner that can't really be created declaratively because of some network interactions etc.? 

Well, generally speaking, once you have the CA up and running, most people won't have to change the config very often. The `step` commands for managing provisioners exist for ease-of-use and automation around the initial setup, not for everyday use.

However, when you're first setting things up, it can be helpful to make changes to the config quickly while you test it out. For example, configuring the OIDC provisioner can be tricky and it's not always going to just work on the first try. But, you can always create a CA in your home directory (that's the default location) and test there.

I would think in this regard step-ca is similar to most other servers. You aren't going into the config all the time to change it.

Reply:
This is somewhat off topic, and if there's a better place for it please let me know...

Since we are talking about immutability and declarative configuration, I'd love to learn more about how NixOS handles TLS certificate renewal in various services.

`step-ca` by default issues short-lived (24 hour) leaf certificates. Depending on the threat model, some CAs might issue certs that only last for five minutes, or less if possible. Not only are they more secure (less attack surface area), they also forces the user to do what's needed to automate renewal. (And [we've spent some time making this easier for people who use systemd](https://smallstep.com/docs/step-ca/certificate-authority-server-production#automated-renewal))

I've spent the last year getting these 24 hour certs to play well with all kinds of services. Usually the service needs to be reloaded when the certificate is updated, but it depends. Every service is a little different. Just yesterday I posted a bug in Keycloak because their Docker container will not reload its certificate files when you restart it. They closed the bug because "containers are immutable". But, that only makes sense if you think certificates and keys are "configuration". They really aren't.

So, it's a real challenge for us and it relates to this question of "how often does config change", not for the CA but for any service that uses TLS certs.

Reply:
As a general rule of thumb, runtime state and secrets are handled separately from Nix itself. If some parts of the configuration relies on them, they would be stored in a known location (e.g., `/var/lib`, `/run/keys`) and be referred to  by path.

Reply:
Hey y'all, I just checked in with the team and discovered an edge case that will impact service unit hardening.

We're [adding support for hardware security modules (HSMs) to `step-ca` right now](https://github.com/smallstep/certificates/pull/457/files). HSM support uses the PKCS#11 software interface (a standard set of C bindings) to access a hardware vendor's HSM. So, `step-ca` will need to link to an arbitrary hardware vendor's PKCS#11 library, and who knows what syscalls and things it will need.

Sometimes a vendor's library will directly control the HSM hardware, sometimes it operates via middleware (eg. with a proxy HTTP API on localhost—YubiHSM2 does this).

That said, most users won't need HSM support. So, we may just leave it unsupported by the service unit for now, and implement it later as a set of overrides that allow hardware access.

Reply:
Thanks for the heads up. I think it's fine to have the default policies a bit strict as long as it's a configuration supported by the upstream project.